wip: fixing pass-by-ref regression in UDO's

hlolli opened the pull request:
starting with having the test actually fail

kunstmusik commented:
Are the assert UDOs not working because of the nature of this bug?

hlolli commented:
@kunstmusik Yes, would fit

hlolli commented:
the target branch wasn't develop so the tests didn't run... anyway, fixing this tonight

hlolli commented:
thinking about it, using UDO to test if UDO works is a bit of anti-pattern :)

vlazzarini commented:
I have a test CSD here to replace that pass by ref test, but I see you've fixed it already.

This is a replacement for udo/test_args_in.csd which should fail if we don't get audio.

```
<CsoundSynthesizer>
<CsOptions>
-n 
</CsOptions>
<CsInstruments>
0dbfs=1
opcode Test(sig:k,fn:i):(a)
a1 flooper2 sig,1,0, 3, 0.1,fn
xout a1
endop

gk1 init 0

i1 ftgen 1, 0, 0, 1, "../fox.wav", 0, 0, 1
instr 1
a1 Test 0.5,1
out a1
gk1 += abs(downsamp(a1))
endin

instr 2
i1 = i(gk1)
if i1 == 0 then
 exitnow(-1)
endif
endin

</CsInstruments>
<CsScore>
i1 0 1
i2 0.5 1
</CsScore>
</CsoundSynthesizer>
```

hlolli commented:
thanks! I'll update it and use it for testing

hlolli committed:
feat(tests): make a false positive test pass_by_ref really fail like it should

hlolli committed:
fix(udo): regressions in perf argument handling

- It was copying i-rate scalar values at init time, which prevented pass-by-reference from working
- It was trying to find parameter names by iterating through the varPool, which doesn't work when multiple UDOs share the same varPool

hlolli commented:
still an issue here, but I want to get review already from coderabbitai even if the tests are not passing yet

@coderabbitai review

hlolli committed:
fix(udo): implement pass-by-reference for i-rate parameters in new-style UDOs

- Fixed parameter name mapping in handle_pass_by_ref() by using xin opcode's output list
- Removed i-rate scalar copying in pass-by-ref mode so opcodes use rewired pointers
- Added safety check in array_copy_value() to skip copying when src == dst
- Guard array aliasing to only apply to new-style UDOs (prevents breaking old-style UDOs)

Fixes pass_by_ref test while maintaining compatibility with test_iarr_operators

hlolli committed:
fix(udo): prevent memory leak when aliasing arrays in pass-by-ref UDOs

Free existing local array allocations (data, sizes) before overwriting with
caller's ARRAYDAT structure to prevent leaking memory on repeated UDO calls.

Applied the fix to both locations:
1. handle_pass_by_ref() for new-style UDO parameters
2. useropcdset() for pass-by-ref array aliasing via xin

hlolli commented on Engine/udo.c:
```diff
@@ -158,52 +161,57 @@ static void handle_pass_by_ref(CSOUND* csound, UOPCODE* p, INSDS* lcurip) {
+                  localString->size = callerString->size;
+                  localString->timestamp = callerString->timestamp;
+                  localString->refcount = -1;
+                }
```
fixed

hlolli committed:
fix(udo): disable pass-by-ref for UDOs with a-rate outputs

Pass-by-reference mode doesn't work correctly for UDOs that have a-rate
outputs because the audio signal path requires proper initialization and
copying that pass-by-ref bypasses. Disabling pass-by-ref for these UDOs
fixes test_args_in.csd while maintaining pass-by-ref optimization for
i-rate and k-rate only UDOs.

Fixes: test_args_in.csd exit code 1 -> 0

kunstmusik commented:
I looked into this and found that that struct arrays PR added a lot of things that broke pass-by-ref (i.e., it added perf time functions that copied values each k-pass, which is the opposite of what we want for pass-by-ref). There's some code I removed locally but there's more to figure out. Will ping @hlolli to discuss further.

vlazzarini commented:
ok, sounds like the thing to do. We can always revert udo.c back to the state before the merge and re-introduce support for structs to it from scratch.

hlolli commented:
I don't want to copy values on each k-rate pass, that's a mistake if that's the case. In fact I tried utmost to design the entire struct members on the contrary. I'll try to find time to fix this today.

vlazzarini commented:
I have not looked at the complete code, but this comment alarmed me:

udo.c:235
```
// Map xin outputs (UDO inputs) - only arrays need pass-by-reference
```

I thought all inputs and outputs were passed by reference in new-style UDOs (if possible). Otherwise all inputs and outputs are passed by copy (old-style and new-style with local sr/ksmps). 

In the previous code `handle_pass_by_ref(CSOUND* csound, UOPCODE* p, INSDS* lcurip)` was *only* called for new-style UDOs, but this seems to have changed (I see a check for `udoinfo->newStyle` in the code now). Asking from a position of ignorance (I don't know the ins and outs of this new code), would it not be better to do as before - only call that function if we need to handle pass by reference? Keeping things separate may simplify and make errors more easily fixable.

hlolli commented:
that comment is incorrect, I was debugging arrays specifically when this was written, the word "only" is a confusion that I didn't fix, you are right it should only be for new style UDOs, good catch I fix it

hlolli committed:
feat: remove 'xout_perf_passByRef' and ensure only new style UDOs use pass_by_ref without copying

kunstmusik commented on Engine/udo.c:
```diff
@@ -411,94 +424,47 @@ int32_t useropcdset(CSOUND *csound, UOPCODE *p)
+  inm->passByRef = buf->opcode_info->newStyle &&
+    parent_ip->ksmps == p->ip->ksmps &&
+    parent_ip->esr == p->ip->esr &&
+    !has_arate_output;  // Disable pass-by-ref for UDOs with a-rate outputs
```
why would you disable pass-by-ref for a-rate outputs? In general the code for pass-by-ref has veered far from what I initially did and I'm doubtful all these exceptions are correct. If anything, they seem to hide a problem rather than solve one.

hlolli replied:
shouldn't be there, has_arate_output shouldn't even exist. Your deletions are there in the patch. I'll see if I can remove the special handling (aka exception)

hlolli replied:
I made a new push that removes this. As for the second part, I think it's better we chat and you show me.

kunstmusik committed:
fix(udo): enhance rewire_argpp to support struct path navigation for pass-by-ref arguments

kunstmusik committed:
fix(udo): corrected xout arg pointer rewiring

kunstmusik committed:
fix(udo): removed unnecessary aliasing

vlazzarini committed:
fixed shortening warning and tightened in arg test for UDO to catch NaNs

vlazzarini committed:
fixed argument count offset

vlazzarini committed:
removed redundant guards

kunstmusik committed:
fix(udo): add test for xout expression evaluation in pass-by-ref UDOs

kunstmusik committed:
Merge branch 'fix/pass-by-ref' of github.com:/csound/csound into fix/pass-by-ref

kunstmusik commented:
@coderabbitai please review

kunstmusik commented:
I think this PR has cleaned up things for pass-by-ref to remove all value copying when the pass-by-ref code paths are used. 

@vlazzarini and @hlolli: Could you review?

@DaveSeidel and @joachimheintz: Could you try pulling this PR branch and testing with the code that was failing for you in #2359 ? 

Want to get as much testing as possible before merging.

DaveSeidel commented:
Yes, I will test it momentarily.

On Sun, Nov 16, 2025 at 12:29 PM Steven Yi ***@***.***> wrote:

> *kunstmusik* left a comment (csound/csound#2362)
> <https://github.com/csound/csound/pull/2362#issuecomment-3538992128>
>
> I think this PR has cleaned up things for pass-by-ref to remove all value
> copying when the pass-by-ref code paths are used.
>
> @vlazzarini <https://github.com/vlazzarini> and @hlolli
> <https://github.com/hlolli>: Could you review?
>
> @DaveSeidel <https://github.com/DaveSeidel> and @joachimheintz
> <https://github.com/joachimheintz>: Could you try pulling this PR branch
> and testing with the code that was failing for you in #2359
> <https://github.com/csound/csound/issues/2359> ?
>
> Want to get as much testing as possible before merging.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/csound/csound/pull/2362#issuecomment-3538992128>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABTVZKW5VTJDPCBANWD23L35CX6LAVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZYHE4TEMJSHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

DaveSeidel commented:
This fix checks out with my code. Thanks, guys! Nice to be able to get
out of detached head mode (which worked fine on my gig last night). :-)

On Sun, Nov 16, 2025 at 12:36 PM Dave Seidel ***@***.***> wrote:

> Yes, I will test it momentarily.
>
> On Sun, Nov 16, 2025 at 12:29 PM Steven Yi ***@***.***>
> wrote:
>
>> *kunstmusik* left a comment (csound/csound#2362)
>> <https://github.com/csound/csound/pull/2362#issuecomment-3538992128>
>>
>> I think this PR has cleaned up things for pass-by-ref to remove all value
>> copying when the pass-by-ref code paths are used.
>>
>> @vlazzarini <https://github.com/vlazzarini> and @hlolli
>> <https://github.com/hlolli>: Could you review?
>>
>> @DaveSeidel <https://github.com/DaveSeidel> and @joachimheintz
>> <https://github.com/joachimheintz>: Could you try pulling this PR branch
>> and testing with the code that was failing for you in #2359
>> <https://github.com/csound/csound/issues/2359> ?
>>
>> Want to get as much testing as possible before merging.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/csound/csound/pull/2362#issuecomment-3538992128>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AABTVZKW5VTJDPCBANWD23L35CX6LAVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZYHE4TEMJSHA>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

vlazzarini commented:
@kunstmusik I had a look earlier this morning after your fixes from yesterday. The fix for xout looked good.

I picked up on a test passing that shouldn't have, tightened it. Then I added my fix for it. 

It all looks ok, but I think we should have a UDO test passing structs in and out if we don't yet have one. We might do already, so this is just in case we do not.

joachimheintz commented:
no good news from my side, unfortunately.
the simple test runs ok, but when i run my complex csd in csoundqt, it 
segfaults after some 50 seconds.
this is the terminal output when i run csoundqt from it:

[D][src/csoundengine.cpp:311 CsoundEngine::keyEventCallback] Released:  49
[D][src/csoundengine.cpp:311 CsoundEngine::keyEventCallback] Released: 
16777249
[D][src/csoundengine.cpp:305 CsoundEngine::keyEventCallback] Pressed:  32
[D][src/csoundengine.cpp:311 CsoundEngine::keyEventCallback] Released:  32
backtrace() returned 17 addresses
/usr/local/lib/libcsound64.so.7.0(+0x7d97a) [0x7eff8647d97a]
/lib/x86_64-linux-gnu/libc.so.6(+0x3fdf0) [0x7eff83449df0]
/lib/x86_64-linux-gnu/libc.so.6(+0x1645f7) [0x7eff8356e5f7]
/usr/local/lib/libcsound64.so.7.0(+0xf379c) [0x7eff864f379c]
/usr/local/lib/libcsound64.so.7.0(+0xb74b1) [0x7eff864b74b1]
/usr/local/lib/libcsound64.so.7.0(+0xb9aa0) [0x7eff864b9aa0]
/usr/local/lib/libcsound64.so.7.0(+0xb9f45) [0x7eff864b9f45]
/usr/local/lib/libcsound64.so.7.0(+0xb87aa) [0x7eff864b87aa]
/usr/local/lib/libcsound64.so.7.0(+0xcf3ad) [0x7eff864cf3ad]
/usr/local/lib/libcsound64.so.7.0(+0xcf937) [0x7eff864cf937]
/usr/local/lib/libcsound64.so.7.0(+0xd0633) [0x7eff864d0633]
/usr/local/lib/libcsound64.so.7.0(csoundPerformKsmps+0xf0) [0x7eff864801fc]
/usr/local/lib/libcsound64.so.7.0(_ZN23CsoundPerformanceThread7PerformEv+0x174) 
[0x7eff8680afe8]
/usr/local/lib/libcsound64.so.7.0(+0x40d703) [0x7eff8680d703]
/usr/local/lib/libcsound64.so.7.0(+0x40b1bc) [0x7eff8680b1bc]
/lib/x86_64-linux-gnu/libc.so.6(+0x92b7b) [0x7eff8349cb7b]
/lib/x86_64-linux-gnu/libc.so.6(+0x1107b8) [0x7eff8351a7b8]
Csound tidy up: Segmentation fault


On 16/11/2025 18:29, Steven Yi wrote:
> *kunstmusik* left a comment (csound/csound#2362) <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3538992128>
> 
> I think this PR has cleaned up things for pass-by-ref to remove all 
> value copying when the pass-by-ref code paths are used.
> 
> @vlazzarini <https://github.com/vlazzarini> and @hlolli <https:// 
> github.com/hlolli>: Could you review?
> 
> @DaveSeidel <https://github.com/DaveSeidel> and @joachimheintz <https:// 
> github.com/joachimheintz>: Could you try pulling this PR branch and 
> testing with the code that was failing for you in #2359 <https:// 
> github.com/csound/csound/issues/2359> ?
> 
> Want to get as much testing as possible before merging.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3538992128>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ 
> AAQYHKWNB7Q67ZT73NQSTSL35CX6LAVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZYHE4TEMJSHA>.
> You are receiving this because you were mentioned.Message ID: <csound/ 
> ***@***.***>
>

kunstmusik commented:
@joachimheintz I tested the code you posted in the #2359:

```
<CsoundSynthesizer>
<CsInstruments>
sr = 44100
ksmps = 32
0dbfs = 1
nchnls = 2

arr@global:k[] = [1,2,3,4,5]

opcode RndInt(low:k,high:k):(k) //max included
  xout(int(random:k(low,high+.9999)))
endop

opcode destr(inarr:k[]):()
    kndx = 0
    while (kndx < lenarray(inarr)) do
        inarr[kndx] = RndInt(-10,-1) ;compare to next line
        ;inarr[kndx] = int(random:k(-10,-1))
        kndx += 1
    od
endop

instr 1
    printarray(arr)
    destr(arr)
    printarray(arr)
    turnoff
endin

</CsInstruments>
<CsScore>
i1 0 1
</CsScore>
</CsoundSynthesizer>

```

and it ran what looked correctly for me:

```
SECTION 1:
new alloc (instance 1) for instr 1:
 1.0000 2.0000 3.0000 4.0000 5.0000 
 -1.0000 -7.0000 -0.0000 -8.0000 -9.0000 
B  0.000 ..  1.000 T  1.000 TT  1.000 M:  0.00000  0.00000
end of Performance
inactive allocs returned to freespace
                   overall amps:  0.00000  0.00000
           overall samples out of range:        0        0
0 errors in performance 
```

I think we might need more help with:

1. If you could create a Debug or RelWithDebInfo build, the crash stack trace should have names for the functions
2. If you could create some minimal example that'd be great. 

The 1st step might be quick to do and get us some good information, but having a reproducible example would of course be great. Given that it's running for an extended period of time before crashing, makes it a harder problem to diagnose without more information.

joachimheintz commented:
yes steven --- i also ran the test and it worked.  that's what i meant 
by "the simple test runs ok".
and i completely understand that it is not really helpful to say "it 
crashed".  but i am sorry: i am in the middle of a big project, and all 
i could do tonight is to test, and then see if the terminal output 
returns something which might be useful for you.
currently i cannot do more; i must go back to the version which worked, 
but i can do a similar test than tonight if there is a new version out 
to fix this issue.

On 16/11/2025 21:36, Steven Yi wrote:
> *kunstmusik* left a comment (csound/csound#2362) <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3539331253>
> 
> @joachimheintz <https://github.com/joachimheintz> I tested the code you 
> posted in the #2359 <https://github.com/csound/csound/issues/2359>:
> 
> |<CsoundSynthesizer> <CsInstruments> sr = 44100 ksmps = 32 0dbfs = 1 
> nchnls = 2 ***@***.***:k[] = [1,2,3,4,5] opcode RndInt(low:k,high:k): 
> (k) //max included xout(int(random:k(low,high+.9999))) endop opcode 
> destr(inarr:k[]):() kndx = 0 while (kndx < lenarray(inarr)) do 
> inarr[kndx] = RndInt(-10,-1) ;compare to next line ;inarr[kndx] = 
> int(random:k(-10,-1)) kndx += 1 od endop instr 1 printarray(arr) 
> destr(arr) printarray(arr) turnoff endin </CsInstruments> <CsScore> i1 0 
> 1 </CsScore> </CsoundSynthesizer> |
> 
> and it ran what looked correctly for me:
> 
> |SECTION 1: new alloc (instance 1) for instr 1: 1.0000 2.0000 3.0000 
> 4.0000 5.0000 -1.0000 -7.0000 -0.0000 -8.0000 -9.0000 B 0.000 .. 1.000 T 
> 1.000 TT 1.000 M: 0.00000 0.00000 end of Performance inactive allocs 
> returned to freespace overall amps: 0.00000 0.00000 overall samples out 
> of range: 0 0 0 errors in performance |
> 
> I think we might need more help with:
> 
>  1. If you could create a Debug or RelWithDebInfo build, the crash stack
>     trace should have names for the functions
>  2. If you could create some minimal example that'd be great.
> 
> The 1st step might be quick to do and get us some good information, but 
> having a reproducible example would of course be great. Given that it's 
> running for an extended period of time before crashing, makes it a 
> harder problem to diagnose without more information.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3539331253>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ 
> AAQYHKVYD5Y4QRD4URJPTM335DN5FAVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZZGMZTCMRVGM>.
> You are receiving this because you were mentioned.Message ID: <csound/ 
> ***@***.***>
>

kunstmusik commented:
Doing a Debug or RelWithDebInfo build is pretty simple, just change CMAKE_BUILD_TYPE in CMakeCache.txt in the build directory and rebuild, then try again. The crash would then have more information. 

For now, I suggest we go ahead and merge this PR as it generally fixes pass-by-reference. We can leave #2359 open and @joachimheintz can follow up when he has time available to investigate further. I have an optimization I want to make as well to pass-by-ref as well as other work that I'd rather have branched from here than what's in develop now. Having this merged into develop would already be an improvement.

@vlazzarini how's that sound to you? (Also, I did add a struct test to tests/commandline/udo/pass_by_ref.csd already so that should address your comment.)

vlazzarini commented:
@kunstmusik I think it's a fair way to proceed, but there is something else I noticed.

With the new code in develop (and this PR) csdtests are running *a lot* slower if compared to a branch before these changes (e.g feature/opcode_aliases).

I tested trapped.csd and figured out why, for develop we have

```
Elapsed time at end of code compilation: real: 0.213s, CPU: 0.213s
```

whereas in the feature/opcode_aliases compilation takes 0.006s. This is a huge slowdown in compilation. Performance is also slower by about 30-40%. In total trapped runs in 1.06s vs 0.5s (new vs old code) (using `time` to test it).

joachimheintz commented:
yes the whole behaviour in csoundqt is very strange.
i cannot stop csound normally, at least not in my program (which 
otherwise had no issues).
thanks for your advice to build as debug, steven; this was indeed easy 
and i can do now.
i am afraid the result is not more meaningful though:

backtrace() returned 11 addresses
/usr/local/lib/libcsound64.so.7.0(+0x7d97a) [0x7fe7f2c7d97a]
/lib/x86_64-linux-gnu/libc.so.6(+0x3fdf0) [0x7fe7efe49df0]
/usr/local/lib/libcsound64.so.7.0(+0xe0112) [0x7fe7f2ce0112]
/usr/local/lib/libcsound64.so.7.0(+0xbc729) [0x7fe7f2cbc729]
/usr/local/lib/libcsound64.so.7.0(+0xbc864) [0x7fe7f2cbc864]
/usr/local/lib/libcsound64.so.7.0(+0xcde62) [0x7fe7f2ccde62]
/usr/local/lib/libcsound64.so.7.0(_ZN23CsoundPerformanceThread7PerformEv+0x24e) 
[0x7fe7f300b940]
/usr/local/lib/libcsound64.so.7.0(+0x40df81) [0x7fe7f300df81]
/usr/local/lib/libcsound64.so.7.0(+0x40ba3a) [0x7fe7f300ba3a]
/lib/x86_64-linux-gnu/libc.so.6(+0x92b7b) [0x7fe7efe9cb7b]
/lib/x86_64-linux-gnu/libc.so.6(+0x1107b8) [0x7fe7eff1a7b8]
Csound tidy up: Segmentation fault


On 16/11/2025 22:31, vlazzarini wrote:
> *vlazzarini* left a comment (csound/csound#2362) <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3539384317>
> 
> @kunstmusik <https://github.com/kunstmusik> I think it's a fair to 
> proceed, but there is something else I noticed.
> 
> With the new code in develop (and this PR) csdtests are running /a lot/ 
> slower if compared to a branch before these changes (e.g feature/ 
> opcode_aliases).
> 
> I tested trapped.csd and figured out why, for develop we have
> 
> |Elapsed time at end of code compilation: real: 0.213s, CPU: 0.213s |
> 
> whereas in the feature/opcode_aliases compilation takes 0.006s. This is 
> a huge slowdown in compilation. Performance is also slower by about 
> 30-40%. In total trapped runs in 1.06s vs 0.5s (new vs old code) (using 
> |time| to test it).
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3539384317>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ 
> AAQYHKR2TPWAZSCIIUOLB5335DUL3AVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZZGM4DIMZRG4>.
> You are receiving this because you were mentioned.Message ID: <csound/ 
> ***@***.***>
>

kunstmusik commented:
@vlazzarini I approved the PR since @hlolli had initialized it.  I will proceed with merging as it sounds like the performance regression is not from this PR. I can run a bisect session in develop to see when the performance issue was introduced given we have a "good" target. 

@joachimheintz I wonder if there was another setting that might have been set (maybe BUILD_RELEASE?) that could have affected symbols not showing up. No worries, we have our hands full with a few things now so we can keep diagnosing in #2359. 

Sidenote: @vlazzarini I have a PR almost ready that parallelizes running the csd tests. It speeds things up quite a bit:

```
16:49:17 - MainThread - INFO - Completed 208/208 tests
16:49:17 - MainThread - INFO - All tests completed in 10.47 seconds
16:49:17 - MainThread - INFO - Parallel execution with 12 workers
16:49:17 - MainThread - INFO - Average time per test: 0.05s
```

vs.

```
16:29:39 - MainThread - INFO - Sequential execution completed in 78.68s
```

I'll have that up shortly.

vlazzarini commented:
> yes the whole behaviour in csoundqt is very strange. i cannot stop csound normally, at least not in my program (which otherwise had no issues). thanks for your advice to build as debug, steven; this was indeed easy and i can do now. i am afraid the result is not more meaningful though: backtrace() returned 11 addresses /usr/local/lib/libcsound64.so.7.0(+0x7d97a) [0x7fe7f2c7d97a] /lib/x86_64-linux-gnu/libc.so.6(+0x3fdf0) [0x7fe7efe49df0] /usr/local/lib/libcsound64.so.7.0(+0xe0112) [0x7fe7f2ce0112] /usr/local/lib/libcsound64.so.7.0(+0xbc729) [0x7fe7f2cbc729] /usr/local/lib/libcsound64.so.7.0(+0xbc864) [0x7fe7f2cbc864] /usr/local/lib/libcsound64.so.7.0(+0xcde62) [0x7fe7f2ccde62] /usr/local/lib/libcsound64.so.7.0(_ZN23CsoundPerformanceThread7PerformEv+0x24e) [0x7fe7f300b940] /usr/local/lib/libcsound64.so.7.0(+0x40df81) [0x7fe7f300df81] /usr/local/lib/libcsound64.so.7.0(+0x40ba3a) [0x7fe7f300ba3a] /lib/x86_64-linux-gnu/libc.so.6(+0x92b7b) [0x7fe7efe9cb7b] /lib/x86_64-linux-gnu/libc.so.6(+0x1107b8) [0x7fe7eff1a7b8] Csound tidy up: Segmentation fault
> […](#)
> On 16/11/2025 22:31, vlazzarini wrote: *vlazzarini* left a comment ([csound/csound#2362](https://github.com/csound/csound/pull/2362)) <https://github.com/ [csound/csound#2362](https://github.com/csound/csound/pull/2362)#issuecomment-3539384317> @kunstmusik <https://github.com/kunstmusik> I think it's a fair to proceed, but there is something else I noticed. With the new code in develop (and this PR) csdtests are running /a lot/ slower if compared to a branch before these changes (e.g feature/ opcode_aliases). I tested trapped.csd and figured out why, for develop we have |Elapsed time at end of code compilation: real: 0.213s, CPU: 0.213s | whereas in the feature/opcode_aliases compilation takes 0.006s. This is a huge slowdown in compilation. Performance is also slower by about 30-40%. In total trapped runs in 1.06s vs 0.5s (new vs old code) (using |time| to test it). — Reply to this email directly, view it on GitHub <https://github.com/ [csound/csound#2362](https://github.com/csound/csound/pull/2362)#issuecomment-3539384317>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ AAQYHKR2TPWAZSCIIUOLB5335DUL3AVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZZGM4DIMZRG4>. You are receiving this because you were mentioned.Message ID: <csound/ ***@***.***>

You could try running csoundqt in gdb, it might give us more info.
Start csoundqt from the terminal under gdb (say the csoundqt binary is called `CsoundQt`)

```
gdb CsoundQt
```

on the gdb prompt, type `run`. When the crash happens you should get a better backtrace,
you can even use the `backtrace` command if you don't get it.

vlazzarini commented:
@kunstmusik yes, for the slowdown definitely not this PR. Luckily trapped.csd seems to be a good test (the old code ran on a par with 6.18 so it is ok), and it does not have any innovation introduced in 7.

Thanks for speeding up csdtests. That's a great help.

vlazzarini commented:
@kunstmusik 

bisect results

```
victor@firebird debug % git bisect bad                    
5c3c6d92879399e029825d832f4c21be68c81706 is the first bad commit
commit 5c3c6d92879399e029825d832f4c21be68c81706 (HEAD)
Author: Hlöðver Sigurðsson <hlolli@gmail.com>
Date:   Wed Oct 15 18:54:18 2025 +0300

    feat: struct arrays memory alignment working

 Engine/auxfd.c                  |   4 +-
 Engine/csound_orc_compile.c     |   2 +-
 Engine/csound_orc_expressions.c | 147 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++--
 Engine/csound_orc_semantics.c   |  40 ++++++++++++++++++++
 Engine/csound_orc_structs.c     |   1 +
 Engine/csound_standard_types.c  |  68 ++++++++++++++++++++++-----------
 Engine/entry.c                  |   2 +-
 Engine/memalloc.c               |  28 +++++++++++++-
 Engine/udo.c                    |  35 ++++++++++++-----
 H/csound_orc_expressions.h      |   1 +
 OOps/array_ops.c                |  59 ++++++++++++++++++++++++++---
 OOps/structs.c                  |  38 +++++++++++++------
 12 files changed, 367 insertions(+), 58 deletions(-)
```

vlazzarini commented:
and I found the culprit - changes in mfree(). Email sent.

joachimheintz commented:
just to let you know that after my project finished (which i ran with 
develop 728c8e4fdc6f1ca30da9577bed651ec79af59057 without any problems) i 
now tried again to run my csd with the current develop, and had similar 
issues (csoundqt freezing).  so i am afraid that still something is 
wrong in the code.
i will try tomorrow to follow victor's advice to run csoundqt with my 
csd in gdb.

On 16/11/2025 22:51, vlazzarini wrote:
> *vlazzarini* left a comment (csound/csound#2362) <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3539399561>
> 
>     yes the whole behaviour in csoundqt is very strange. i cannot stop
>     csound normally, at least not in my program (which otherwise had no
>     issues). thanks for your advice to build as debug, steven; this was
>     indeed easy and i can do now. i am afraid the result is not more
>     meaningful though: backtrace() returned 11 addresses /usr/local/lib/
>     libcsound64.so.7.0(+0x7d97a) [0x7fe7f2c7d97a] /lib/x86_64-linux-gnu/
>     libc.so.6(+0x3fdf0) [0x7fe7efe49df0] /usr/local/lib/
>     libcsound64.so.7.0(+0xe0112) [0x7fe7f2ce0112] /usr/local/lib/
>     libcsound64.so.7.0(+0xbc729) [0x7fe7f2cbc729] /usr/local/lib/
>     libcsound64.so.7.0(+0xbc864) [0x7fe7f2cbc864] /usr/local/lib/
>     libcsound64.so.7.0(+0xcde62) [0x7fe7f2ccde62] /usr/local/lib/
>     libcsound64.so.7.0(_ZN23CsoundPerformanceThread7PerformEv+0x24e)
>     [0x7fe7f300b940] /usr/local/lib/libcsound64.so.7.0(+0x40df81)
>     [0x7fe7f300df81] /usr/local/lib/libcsound64.so.7.0(+0x40ba3a)
>     [0x7fe7f300ba3a] /lib/x86_64-linux-gnu/libc.so.6(+0x92b7b)
>     [0x7fe7efe9cb7b] /lib/x86_64-linux-gnu/libc.so.6(+0x1107b8)
>     [0x7fe7eff1a7b8] Csound tidy up: Segmentation fault
>     … <#>
>     On 16/11/2025 22:31, vlazzarini wrote: /vlazzarini/ left a comment
>     (csound/csound#2362 <https://github.com/csound/csound/pull/2362>)
>     <https://github.com/ <https://github.com/> csound/csound#2362
>     <https://github.com/csound/csound/
>     pull/2362>#issuecomment-3539384317> @kunstmusik <https://github.com/
>     kunstmusik> https://github.com/kunstmusik <https://github.com/
>     kunstmusik> I think it's a fair to proceed, but there is something
>     else I noticed. With the new code in develop (and this PR) csdtests
>     are running /a lot/ slower if compared to a branch before these
>     changes (e.g feature/ opcode_aliases). I tested trapped.csd and
>     figured out why, for develop we have |Elapsed time at end of code
>     compilation: real: 0.213s, CPU: 0.213s | whereas in the feature/
>     opcode_aliases compilation takes 0.006s. This is a huge slowdown in
>     compilation. Performance is also slower by about 30-40%. In total
>     trapped runs in 1.06s vs 0.5s (new vs old code) (using |time| to
>     test it). — Reply to this email directly, view it on GitHub
>     <https://github.com/ <https://github.com/> csound/csound#2362
>     <https://github.com/csound/csound/
>     pull/2362>#issuecomment-3539384317>, or unsubscribe <https://
>     github.com/notifications/unsubscribe-auth/ <https://github.com/
>     notifications/unsubscribe-auth/>
>     AAQYHKR2TPWAZSCIIUOLB5335DUL3AVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZZGM4DIMZRG4>. You are receiving this because you were mentioned.Message ID: <csound/ /*@*/.***>
> 
> You could try running csoundqt in gdb, it might give us more info.
> Start csoundqt from the terminal under gdb (say the csoundqt binary is 
> called |CsoundQt|)
> 
> |gdb CsoundQt |
> 
> on the gdb prompt, type |run|. When the crash happens you should get a 
> better backtrace,
> you can even use the |backtrace| command if you don't get it.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3539399561>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ 
> AAQYHKS27ISZCZI6T7SIRFL35DWVTAVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZZGM4TSNJWGE>.
> You are receiving this because you were mentioned.Message ID: <csound/ 
> ***@***.***>
>

joachimheintz commented:
this is now the output of gdb:

[New Thread 0x7fffd82f36c0 (LWP 22396)]
[New Thread 0x7fffd810c6c0 (LWP 22397)]
Cannot lock down 107341340 byte memory area (Cannot allocate memory)
[New Thread 0x7fffd808b6c0 (LWP 22398)]
Cannot use real-time scheduling (RR/5) (1: Operation not permitted)
JackClient::AcquireSelfRealTime error
[New Thread 0x7fffbd9026c0 (LWP 22399)]
[Thread 0x7fffbd9026c0 (LWP 22399) exited]
[New Thread 0x7fffbd9026c0 (LWP 22400)]
[Thread 0x7fffbd9026c0 (LWP 22400) exited]
[New Thread 0x7fffbd9026c0 (LWP 22401)]
[Thread 0x7fffbd9026c0 (LWP 22401) exited]
[New Thread 0x7fffbd9026c0 (LWP 22402)]
[Thread 0x7fffbd9026c0 (LWP 22402) exited]
[New Thread 0x7fffbd8816c0 (LWP 22403)]
[D][src/csoundengine.cpp:311 CsoundEngine::keyEventCallback] Released:  49
[D][src/csoundengine.cpp:311 CsoundEngine::keyEventCallback] Released: 
16777249
[D][src/csoundengine.cpp:305 CsoundEngine::keyEventCallback] Pressed:  32
[D][src/csoundengine.cpp:311 CsoundEngine::keyEventCallback] Released:  32
free(): invalid pointer

Thread 25 "CsoundQt-d-cs7-" received signal SIGABRT, Aborted.
[Switching to Thread 0x7fffbd8816c0 (LWP 22403)]
0x00007ffff4a9e95c in ?? () from /lib/x86_64-linux-gnu/libc.so.6
(gdb)


On 01/12/2025 20:16, joachimheintz wrote:
> *joachimheintz* left a comment (csound/csound#2362) <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3598449584>
> just to let you know that after my project finished (which i ran with
> develop 728c8e4fdc6f1ca30da9577bed651ec79af59057 without any problems) i
> now tried again to run my csd with the current develop, and had similar
> issues (csoundqt freezing). so i am afraid that still something is
> wrong in the code.
> i will try tomorrow to follow victor's advice to run csoundqt with my
> csd in gdb.
> 
> On 16/11/2025 22:51, vlazzarini wrote:
>  > *vlazzarini* left a comment (csound/csound#2362) <https://github.com/
>  > csound/csound/pull/2362#issuecomment-3539399561>
>  >
>  > yes the whole behaviour in csoundqt is very strange. i cannot stop
>  > csound normally, at least not in my program (which otherwise had no
>  > issues). thanks for your advice to build as debug, steven; this was
>  > indeed easy and i can do now. i am afraid the result is not more
>  > meaningful though: backtrace() returned 11 addresses /usr/local/lib/
>  > libcsound64.so.7.0(+0x7d97a) [0x7fe7f2c7d97a] /lib/x86_64-linux-gnu/
>  > libc.so.6(+0x3fdf0) [0x7fe7efe49df0] /usr/local/lib/
>  > libcsound64.so.7.0(+0xe0112) [0x7fe7f2ce0112] /usr/local/lib/
>  > libcsound64.so.7.0(+0xbc729) [0x7fe7f2cbc729] /usr/local/lib/
>  > libcsound64.so.7.0(+0xbc864) [0x7fe7f2cbc864] /usr/local/lib/
>  > libcsound64.so.7.0(+0xcde62) [0x7fe7f2ccde62] /usr/local/lib/
>  > libcsound64.so.7.0(_ZN23CsoundPerformanceThread7PerformEv+0x24e)
>  > [0x7fe7f300b940] /usr/local/lib/libcsound64.so.7.0(+0x40df81)
>  > [0x7fe7f300df81] /usr/local/lib/libcsound64.so.7.0(+0x40ba3a)
>  > [0x7fe7f300ba3a] /lib/x86_64-linux-gnu/libc.so.6(+0x92b7b)
>  > [0x7fe7efe9cb7b] /lib/x86_64-linux-gnu/libc.so.6(+0x1107b8)
>  > [0x7fe7eff1a7b8] Csound tidy up: Segmentation fault
>  > … <#>
>  > On 16/11/2025 22:31, vlazzarini wrote: /vlazzarini/ left a comment
>  > (csound/csound#2362 <https://github.com/csound/csound/pull/2362>)
>  > <https://github.com/ <https://github.com/> csound/csound#2362
>  > <https://github.com/csound/csound/
>  > pull/2362>#issuecomment-3539384317> @kunstmusik <https://github.com/
>  > kunstmusik> https://github.com/kunstmusik <https://github.com/
>  > kunstmusik> I think it's a fair to proceed, but there is something
>  > else I noticed. With the new code in develop (and this PR) csdtests
>  > are running /a lot/ slower if compared to a branch before these
>  > changes (e.g feature/ opcode_aliases). I tested trapped.csd and
>  > figured out why, for develop we have |Elapsed time at end of code
>  > compilation: real: 0.213s, CPU: 0.213s | whereas in the feature/
>  > opcode_aliases compilation takes 0.006s. This is a huge slowdown in
>  > compilation. Performance is also slower by about 30-40%. In total
>  > trapped runs in 1.06s vs 0.5s (new vs old code) (using |time| to
>  > test it). — Reply to this email directly, view it on GitHub
>  > <https://github.com/ <https://github.com/> csound/csound#2362
>  > <https://github.com/csound/csound/
>  > pull/2362>#issuecomment-3539384317>, or unsubscribe <https://
>  > github.com/notifications/unsubscribe-auth/ <https://github.com/
>  > notifications/unsubscribe-auth/>
>  > 
> AAQYHKR2TPWAZSCIIUOLB5335DUL3AVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZZGM4DIMZRG4>. You are receiving this because you were mentioned.Message ID: <csound/ /*@*/.***>
>  >
>  > You could try running csoundqt in gdb, it might give us more info.
>  > Start csoundqt from the terminal under gdb (say the csoundqt binary is
>  > called |CsoundQt|)
>  >
>  > |gdb CsoundQt |
>  >
>  > on the gdb prompt, type |run|. When the crash happens you should get a
>  > better backtrace,
>  > you can even use the |backtrace| command if you don't get it.
>  >
>  > —
>  > Reply to this email directly, view it on GitHub <https://github.com/
>  > csound/csound/pull/2362#issuecomment-3539399561>, or unsubscribe
>  > <https://github.com/notifications/unsubscribe-auth/
>  > 
> AAQYHKS27ISZCZI6T7SIRFL35DWVTAVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKMZZGM4TSNJWGE>.
>  > You are receiving this because you were mentioned.Message ID: <csound/
>  > ***@***.***>
>  >
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3598449584>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ 
> AAQYHKWYSSYQRCTGYXJSB6D37SHXVAVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKOJYGQ2DSNJYGQ>.
> You are receiving this because you are subscribed to this thread.Message 
> ID: ***@***.***>
>

vlazzarini commented:
Can you get a backtrace? Use the backtrace command in gdb.

joachimheintz commented:
yes:

[New Thread 0x7fffd82da6c0 (LWP 23246)]
[New Thread 0x7fffd80f36c0 (LWP 23247)]
Cannot lock down 107341340 byte memory area (Cannot allocate memory)
[New Thread 0x7fffbd9026c0 (LWP 23248)]
Cannot use real-time scheduling (RR/5) (1: Operation not permitted)
JackClient::AcquireSelfRealTime error
[New Thread 0x7fffbd8816c0 (LWP 23249)]
[Thread 0x7fffbd8816c0 (LWP 23249) exited]
[New Thread 0x7fffbd8816c0 (LWP 23250)]
[Thread 0x7fffbd8816c0 (LWP 23250) exited]
[New Thread 0x7fffbd8816c0 (LWP 23251)]
[Thread 0x7fffbd8816c0 (LWP 23251) exited]
[New Thread 0x7fffbd8816c0 (LWP 23252)]
[Thread 0x7fffbd8816c0 (LWP 23252) exited]
[New Thread 0x7fffbd8006c0 (LWP 23253)]
[D][src/csoundengine.cpp:305 CsoundEngine::keyEventCallback] Pressed:  32
[D][src/csoundengine.cpp:311 CsoundEngine::keyEventCallback] Released:  32
free(): invalid pointer

Thread 25 "CsoundQt-d-cs7-" received signal SIGABRT, Aborted.
[Switching to Thread 0x7fffbd8006c0 (LWP 23253)]
0x00007ffff4a9e95c in ?? () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) backtrace
#0  0x00007ffff4a9e95c in ??? () at /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007ffff4a49cc2 in raise () at /lib/x86_64-linux-gnu/libc.so.6
#2  0x00007ffff4a324ac in abort () at /lib/x86_64-linux-gnu/libc.so.6
#3  0x00007ffff4a33291 in ??? () at /lib/x86_64-linux-gnu/libc.so.6
#4  0x00007ffff4aa8465 in ??? () at /lib/x86_64-linux-gnu/libc.so.6
#5  0x00007ffff4aad3dc in free () at /lib/x86_64-linux-gnu/libc.so.6
#6  0x00007ffff7acc7f9 in mfree () at /usr/local/lib/libcsound64.so.7.0
#7  0x00007ffff7ae1cc5 in array_copy_value ()
     at /usr/local/lib/libcsound64.so.7.0
#8  0x00007ffff7ac5dc3 in xoutset () at /usr/local/lib/libcsound64.so.7.0
#9  0x00007ffff7ac4f23 in useropcdset () at 
/usr/local/lib/libcsound64.so.7.0
#10 0x00007ffff7aba541 in init_pass () at /usr/local/lib/libcsound64.so.7.0
#11 0x00007ffff7abcb13 in insert_new () at /usr/local/lib/libcsound64.so.7.0
#12 0x00007ffff7abcfb8 in insert () at /usr/local/lib/libcsound64.so.7.0
#13 0x00007ffff7abb83a in insert_event () at 
/usr/local/lib/libcsound64.so.7.0
#14 0x00007ffff7ad2357 in process_score_event ()
     at /usr/local/lib/libcsound64.so.7.0
#15 0x00007ffff7ad28e1 in process_rt_event ()
     at /usr/local/lib/libcsound64.so.7.0
#16 0x00007ffff7ad35dd in sense_events () at 
/usr/local/lib/libcsound64.so.7.0
#17 0x00007ffff7a831fc in csoundPerformKsmps ()
     at /usr/local/lib/libcsound64.so.7.0
#18 0x00007ffff7e0ef66 in CsoundPerformanceThread::Perform() ()
--Type <RET> for more, q to quit, c to continue without paging--
     at /usr/local/lib/libcsound64.so.7.0
#19 0x00007ffff7e11681 in CsPerfThread_PerformScore::Perform() ()
     at /usr/local/lib/libcsound64.so.7.0
#20 0x00007ffff7e0f13a in csoundPerformanceThread_ ()
     at /usr/local/lib/libcsound64.so.7.0
#21 0x00007ffff4a9cb7b in ??? () at /lib/x86_64-linux-gnu/libc.so.6
#22 0x00007ffff4b1a7b8 in ??? () at /lib/x86_64-linux-gnu/libc.so.6


On 03/12/2025 18:26, vlazzarini wrote:
> *vlazzarini* left a comment (csound/csound#2362) <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3607968482>
> 
> Can you get a backtrace? Use the backtrace command in gdb.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3607968482>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ 
> AAQYHKSC7GPGXE7YBRHRMYL374MKXAVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMMBXHE3DQNBYGI>.
> You are receiving this because you were mentioned.Message ID: <csound/ 
> ***@***.***>
>

vlazzarini commented:
Right: in your code are you using classic-style opcodes or local ksmps or local sr at all anywhere?

The code where the crash happens should not be reached by pass-by-ref (newstyle udos without local ksmps or local sr)

joachimheintz commented:
not sure i understand.
do you mean classic-style UDOs, or do you mean classic-style coding?
i don't think i use any local ksmsps or local sr.

On 03/12/2025 20:30, vlazzarini wrote:
> *vlazzarini* left a comment (csound/csound#2362) <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3608500222>
> 
> Right: in your code are you using classic-style opcodes or local ksmps 
> or local sr at all anywhere?
> 
> The code where the crash happens should not be reached by pass-by-ref 
> (newstyle udos without local ksmps or local sr)
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3608500222>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ 
> AAQYHKRQ3APMRTBM4MPDUXL37426RAVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMMBYGUYDAMRSGI>.
> You are receiving this because you were mentioned.Message ID: <csound/ 
> ***@***.***>
>

vlazzarini commented:
I mean udos, not newstyle, but the original form, pass by copy.

joachimheintz commented:
yes i use both styles in this csd.
if you want me to, i can send you the code via email.

On 03/12/2025 20:50, vlazzarini wrote:
> *vlazzarini* left a comment (csound/csound#2362) <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3608571714>
> 
> I mean udos, not newstyle, but the original form, pass by copy.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3608571714>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ 
> AAQYHKW2U4ICKPSR4SBYXS33745G5AVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMMBYGU3TCNZRGQ>.
> You are receiving this because you were mentioned.Message ID: <csound/ 
> ***@***.***>
>

vlazzarini commented:
So the problem is in one of the oldstyle UDOs that produces an i-array as an output. Do you have many of those?

vlazzarini commented:
I see now that the code that is crashing was changed from before. I'll open up a new PR to revert that and see if it fixes your problem.

joachimheintz commented:
yes i see three of them.

On 03/12/2025 21:01, vlazzarini wrote:
> *vlazzarini* left a comment (csound/csound#2362) <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3608613034>
> 
> So the problem is in one of the oldstyle UDOs that produces an i-array 
> as an output. Do you have many of those?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ 
> csound/csound/pull/2362#issuecomment-3608613034>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ 
> AAQYHKWD7ITKGLIIQRIY3Q33746TNAVCNFSM6AAAAACLWARK5CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMMBYGYYTGMBTGQ>.
> You are receiving this because you were mentioned.Message ID: <csound/ 
> ***@***.***>
>

vlazzarini commented:
Ok, here's the PR: #2378
Could you switch to this branch and test it?

vlazzarini commented:
I can reproduce the crash here now - fixed in that PR
